File: java/src/main/java/com/sketches_oxide/quantiles/DDSketch.java
package com.sketches_oxide.quantiles;

import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.MergeableSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * DDSketch - Quantile Estimation with Relative Error Guarantees.
 *
 * A quantile sketch that provides relative error guarantees on the quantile values
 * themselves, rather than on ranks. This makes it particularly suitable for metrics
 * where the magnitude of values varies greatly (e.g., latency distributions).
 *
 * <p><strong>Algorithm:</strong>
 * <ul>
 *   <li>Maps values to buckets using a logarithmic mapping</li>
 *   <li>Bucket boundaries grow geometrically</li>
 *   <li>Relative accuracy: for any quantile q, the returned value v satisfies
 *       |v - true_value| &lt;= alpha * true_value</li>
 *   <li>Space: O(log(max/min) / log(1 + alpha)) buckets</li>
 * </ul>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>Latency monitoring (p50, p95, p99)</li>
 *   <li>SLO/SLA compliance checking</li>
 *   <li>Performance benchmarking</li>
 *   <li>Any metric where relative error is more meaningful than absolute</li>
 * </ul>
 *
 * <p><strong>Relative vs Absolute Error:</strong>
 * <ul>
 *   <li>Relative: "within 1% of the true value" - good for varying magnitudes</li>
 *   <li>Absolute: "within rank epsilon" - good for uniform distributions</li>
 *   <li>DDSketch is ideal when you care about percentage accuracy</li>
 * </ul>
 *
 * <p><strong>Example:</strong>
 * <pre>
 * // Monitor API latencies with 1% relative accuracy
 * try (DDSketch sketch = new DDSketch(0.01)) {
 *     // Record latencies in milliseconds
 *     for (double latency : latencies) {
 *         sketch.update(latency);
 *     }
 *
 *     // Get percentiles
 *     System.out.println("p50: " + sketch.quantile(0.5) + "ms");
 *     System.out.println("p95: " + sketch.quantile(0.95) + "ms");
 *     System.out.println("p99: " + sketch.quantile(0.99) + "ms");
 *
 *     // Get multiple quantiles efficiently
 *     double[] percentiles = sketch.quantiles(new double[]{0.5, 0.9, 0.95, 0.99});
 * }
 * </pre>
 *
 * @see <a href="https://arxiv.org/abs/1908.10693">DDSketch Paper</a>
 */
public final class DDSketch extends NativeSketch implements MergeableSketch<DDSketch> {

    private final double relativeAccuracy;

    /**
     * Create a new DDSketch.
     *
     * @param relativeAccuracy the relative accuracy parameter alpha (0 &lt; alpha &lt; 1)
     *                         Smaller values give more accurate results but use more space.
     *                         Typical values: 0.01 (1%), 0.001 (0.1%)
     * @throws IllegalArgumentException if relativeAccuracy not in (0, 1)
     * @throws RuntimeException if native memory allocation fails
     */
    public DDSketch(double relativeAccuracy) {
        if (relativeAccuracy <= 0 || relativeAccuracy >= 1) {
            throw new IllegalArgumentException(
                    "relativeAccuracy must be in range (0, 1), got: " + relativeAccuracy);
        }

        this.relativeAccuracy = relativeAccuracy;
        this.nativePtr = SketchOxideNative.ddsketch_new(relativeAccuracy);

        if (this.nativePtr == 0) {
            throw new RuntimeException("Failed to allocate DDSketch");
        }
    }

    /**
     * Private constructor for deserialization.
     */
    private DDSketch(long nativePtr, double relativeAccuracy) {
        this.nativePtr = nativePtr;
        this.relativeAccuracy = relativeAccuracy;
    }

    /**
     * Add a value to the sketch.
     *
     * @param value the value to add (must be positive for logarithmic mapping)
     * @throws IllegalStateException if the sketch has been closed
     */
    public void update(double value) {
        checkAlive();
        SketchOxideNative.ddsketch_update(nativePtr, value);
    }

    /**
     * Get the estimated value at a given quantile.
     *
     * The returned value v satisfies: |v - true_value| &lt;= alpha * true_value,
     * where alpha is the relativeAccuracy parameter.
     *
     * @param q the quantile to query (0 &lt;= q &lt;= 1)
     *          0.5 = median, 0.95 = 95th percentile, etc.
     * @return the estimated value at quantile q
     * @throws IllegalStateException if the sketch has been closed
     * @throws IllegalArgumentException if q not in [0, 1]
     */
    public double quantile(double q) {
        checkAlive();
        if (q < 0 || q > 1) {
            throw new IllegalArgumentException("quantile must be in [0, 1], got: " + q);
        }
        return SketchOxideNative.ddsketch_quantile(nativePtr, q);
    }

    /**
     * Get estimated values at multiple quantiles efficiently.
     *
     * More efficient than calling quantile() multiple times as it
     * processes all queries in a single pass.
     *
     * @param quantiles array of quantiles to query (each in [0, 1])
     * @return array of estimated values, one per input quantile
     * @throws IllegalStateException if the sketch has been closed
     * @throws IllegalArgumentException if any quantile not in [0, 1]
     * @throws NullPointerException if quantiles is null
     */
    public double[] quantiles(double[] quantiles) {
        checkAlive();
        if (quantiles == null) {
            throw new NullPointerException("quantiles cannot be null");
        }

        double[] results = new double[quantiles.length];
        for (int i = 0; i < quantiles.length; i++) {
            if (quantiles[i] < 0 || quantiles[i] > 1) {
                throw new IllegalArgumentException(
                        "quantile must be in [0, 1], got: " + quantiles[i] + " at index " + i);
            }
            results[i] = SketchOxideNative.ddsketch_quantile(nativePtr, quantiles[i]);
        }
        return results;
    }

    /**
     * Merge another DDSketch into this one.
     *
     * After merging, this sketch represents the combined distribution.
     * Both sketches must have the same relativeAccuracy parameter.
     *
     * @param other the sketch to merge into this one
     * @throws IllegalStateException if either sketch is closed
     * @throws IllegalArgumentException if relativeAccuracy doesn't match
     * @throws NullPointerException if other is null
     */
    @Override
    public void merge(DDSketch other) {
        checkAlive();
        if (other == null) {
            throw new NullPointerException("other cannot be null");
        }
        other.checkAlive();

        if (Math.abs(this.relativeAccuracy - other.relativeAccuracy) > 1e-9) {
            throw new IllegalArgumentException(
                    "Cannot merge DDSketches with different relativeAccuracy: " +
                            this.relativeAccuracy + " vs " + other.relativeAccuracy);
        }

        SketchOxideNative.ddsketch_merge(this.nativePtr, other.nativePtr);
    }

    /**
     * Serialize the sketch to binary format.
     *
     * @return binary representation of the sketch
     * @throws IllegalStateException if the sketch has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.ddsketch_serialize(nativePtr);
    }

    /**
     * Deserialize a DDSketch from binary format.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new DDSketch instance
     * @throws IllegalArgumentException if data is invalid or corrupt
     * @throws NullPointerException if data is null
     */
    public static DDSketch deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.ddsketch_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized DDSketch data");
        }

        return new DDSketch(ptr, 0.01);
    }

    /**
     * Get the relative accuracy parameter.
     *
     * @return the relativeAccuracy used at creation
     */
    public double getRelativeAccuracy() {
        return relativeAccuracy;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.ddsketch_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        if (nativePtr != 0) {
            return String.format("DDSketch(relativeAccuracy=%.4f)", relativeAccuracy);
        }
        return "DDSketch(closed)";
    }
}

File: dotnet/SketchOxide/src/Quantiles/DDSketch.cs
using System;
using SketchOxide.Native;

namespace SketchOxide.Quantiles;

/// <summary>
/// DDSketch quantile estimator with relative error guarantees.
///
/// Provides percentile estimation where the error is relative to the actual value,
/// making it particularly suitable for latency measurements and other metrics
/// where relative accuracy is more important than absolute accuracy.
/// </summary>
/// <remarks>
/// DDSketch guarantees that for any quantile q, the returned value v satisfies:
/// true_value * (1 - relative_accuracy) &lt;= v &lt;= true_value * (1 + relative_accuracy)
/// </remarks>
public sealed class DDSketch : NativeSketch
{
    private readonly double _relativeAccuracy;

    /// <summary>
    /// Creates a new DDSketch with the specified relative accuracy.
    /// </summary>
    /// <param name="relativeAccuracy">Relative accuracy in range (0, 1).
    /// Smaller values give better accuracy but use more space.
    /// A value of 0.01 means 1% relative error.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if relativeAccuracy is outside (0, 1).</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public DDSketch(double relativeAccuracy)
    {
        if (relativeAccuracy <= 0 || relativeAccuracy >= 1)
            throw new ArgumentOutOfRangeException(nameof(relativeAccuracy), relativeAccuracy, "Relative accuracy must be in range (0, 1)");

        _relativeAccuracy = relativeAccuracy;
        NativePtr = SketchOxideNative.ddsketch_new(relativeAccuracy);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native DDSketch");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private DDSketch(double relativeAccuracy, nuint ptr)
    {
        _relativeAccuracy = relativeAccuracy;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the relative accuracy parameter of this sketch.
    /// </summary>
    public double RelativeAccuracy
    {
        get
        {
            CheckAlive();
            return _relativeAccuracy;
        }
    }

    /// <summary>
    /// Updates the sketch with a new value.
    /// </summary>
    /// <param name="value">The value to add to the sketch.</param>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public void Update(double value)
    {
        CheckAlive();
        SketchOxideNative.ddsketch_update(NativePtr, value);
    }

    /// <summary>
    /// Returns the estimated value at the given quantile.
    /// </summary>
    /// <param name="q">The quantile to query, in range [0, 1].
    /// 0 = minimum, 0.5 = median, 1 = maximum.</param>
    /// <returns>The estimated value at quantile q.</returns>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if q is outside [0, 1].</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Quantile(double q)
    {
        CheckAlive();
        if (q < 0 || q > 1)
            throw new ArgumentOutOfRangeException(nameof(q), q, "Quantile must be in range [0, 1]");

        return SketchOxideNative.ddsketch_quantile(NativePtr, q);
    }

    /// <summary>
    /// Gets the estimated minimum value (p0).
    /// </summary>
    /// <returns>The estimated minimum value.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Min() => Quantile(0);

    /// <summary>
    /// Gets the estimated maximum value (p100).
    /// </summary>
    /// <returns>The estimated maximum value.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Max() => Quantile(1);

    /// <summary>
    /// Gets the estimated median value (p50).
    /// </summary>
    /// <returns>The estimated median value.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Median() => Quantile(0.5);

    /// <summary>
    /// Serializes the sketch to a byte array.
    /// </summary>
    /// <returns>Serialized sketch bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.ddsketch_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a DDSketch from a byte array.
    /// </summary>
    /// <param name="data">Serialized sketch bytes.</param>
    /// <param name="relativeAccuracy">The relative accuracy parameter used when creating the original sketch.</param>
    /// <returns>A new DDSketch instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static DDSketch Deserialize(byte[] data, double relativeAccuracy)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.ddsketch_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize DDSketch: invalid data");

        return new DDSketch(relativeAccuracy, ptr);
    }

    /// <summary>
    /// Returns a string representation of the sketch.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "DDSketch(disposed)";
        return $"DDSketch(relativeAccuracy={_relativeAccuracy})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.ddsketch_free(NativePtr);
            NativePtr = 0;
        }
    }
}
